In [1]:
import pandas as pd

df = pd.read_csv("dataset_with_explanations_full.csv")

In [7]:
df.shape

(20147, 9)

In [4]:
from sklearn.model_selection import train_test_split

# Split the data ensuring stratification based on the 'label' column
train_df, test_df = train_test_split(df, test_size=0.1, stratify=df['label'], random_state=42)

# Verify the label distribution in both sets
print("Train label distribution:\n", train_df['label'].value_counts(normalize=True))
print("Test label distribution:\n", test_df['label'].value_counts(normalize=True))


Train label distribution:
 label
normal        0.404644
hatespeech    0.309398
offensive     0.285959
Name: proportion, dtype: float64
Test label distribution:
 label
normal        0.404467
hatespeech    0.309677
offensive     0.285856
Name: proportion, dtype: float64


(2015, 9)

**Lets Create test df in a way that, we get same number of rows for each class**

**since we have 20147 of total rows. we get 2015 for 10% of it. and then we get 672 for each class**

In [6]:
import numpy as np

# Define the number of samples you want for each class in the test set
num_test_samples_per_class = 672

# Create an empty list to store the test samples
test_samples = []

# For each class, sample the required number of examples from the dataframe
for label in df['label'].unique():
    # Filter rows that belong to the current class
    class_df = df[df['label'] == label]
    
    # Sample the specified number of rows for the test set
    class_test_df = class_df.sample(n=num_test_samples_per_class, random_state=42)
    
    # Append the sampled rows to the test_samples list
    test_samples.append(class_test_df)

# Concatenate the test samples into a single DataFrame
test_df = pd.concat(test_samples)

# Drop the selected test samples from the original dataframe to create the train set
train_df = df.drop(test_df.index)

# Verify the distribution in the train and test sets
print("Train label distribution:\n", train_df['label'].value_counts(normalize=True))
print("Test label distribution:\n", test_df['label'].value_counts(normalize=True))


Train label distribution:
 label
normal        0.412553
hatespeech    0.306767
offensive     0.280679
Name: proportion, dtype: float64
Test label distribution:
 label
normal        0.333333
hatespeech    0.333333
offensive     0.333333
Name: proportion, dtype: float64


In [10]:
test_df.to_csv("TEST_DF.csv")

In [11]:
train_df.to_csv("TRAIN_DF.csv")